### PR TITLE
[WIP] fix(required-parent): validate parent nodes for desired context role

### PR DESCRIPTION
**Pull request by jkodu**
Closes issue:
- https://github.com/dequelabs/axe-core/issues/2012

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**

- [ ] Follows the commit message policy, appropriate for next version
- [ ] Code is reviewed for security

### CODE REVIEW

File: lib/checks/aria/required-parent.js [hunk 1]
```diff
@@ -62,26 +79,26 @@ function getAriaOwners(element) {
...
-	return true;
-}
+/**
+ * Recursively attempt to find parent with a desired role
+ * @param {HTMLElement} elm element
+ * @param {String} role aria role
+ * @returns {Boolean}
+ */
+function hasParentWithRole(elm, role) {
+	const parent = dom.getComposedParent(elm);
```
Comment (straker) [inline lib/checks/aria/required-parent.js]: Wouldn't it be easier to pass `virtualNode` into the function? Then you wouldn't have to call `getComposedParent` and could instead use `virtalNode.parent` which is already the composed parent.

File: lib/checks/aria/required-parent.js [hunk 2]
```diff
@@ -62,26 +79,26 @@ function getAriaOwners(element) {
...
+	if (!parent) {
+		return false;
+	}
 
-var owners = getAriaOwners(node);
+	const parentRole = aria.getRole(parent, { noImplicit: true });
```
Comment (straker) [inline lib/checks/aria/required-parent.js]: Implicit roles should probably be considered. Current axe does not fail this:

```
<ul>
  <div role="listitem">List item 1</div>
</ul>
```

but these changes will.

**Comment by straker** [comment]
Could you add a test ensuring this passes so we don't get regressions?

```html
<ul>
  <div role="listitem">List item 1</div>
</ul>
```

File: lib/checks/aria/required-parent.js [hunk 3]
```diff
@@ -1,5 +1,29 @@
+const { aria, dom } = axe.commons;
```
Comment (WilcoFiers) [inline lib/checks/aria/required-parent.js]: I'd really like us to clean this up... my initial request for this was to create an `aria.getOwnerVirtual()`. I still think we should do this. Here's what this check would look like if we do this, which I think is far easier to read then what we have now:

```js
const role = aria.getRole(virtualNode)
const requiredContext = aria.requiredContext(role);
if (!requiredContext) {
  return true
}

// Accounts for role=presentation, as well as aria-owns
const owner = aria.getOwnerVirtual(virtualNode)
if (!owner) {
  return undefined;
}

const ownerRole = aria.getRole(owner)
if (requiredContext.some((contextRole) => contextRole === ownerRole)) {
  return true;
} else {
  this.relatedNode(owner);
  this.data(requiredContext);
  return false
}
```

Comment (jkodu) [reply lib/checks/aria/required-parent.js]: I would really like to keep this PR scope to fix the bug only & not go down the route of introducing nice to have's and helper function.

I will look into what the `getOwnerVirtual` entails.

File: lib/commons/aria/get-owner-virtual.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * @returns {HTMLElement | undefined}
+ */
+function getAriaOwner(vNode) {
+	if (!vNode) {
+		return;
+	}
+
+	const id = vNode.attr('id');
+	if (!id) {
+		return getAriaOwner(vNode.parent);
```
Comment (WilcoFiers) [inline lib/commons/aria/get-owner-virtual.js]: aria-owns doesn't inherit descendants. This should just be:
```suggestion
		return;
```

File: lib/commons/aria/get-owner-virtual.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+	const doc = dom.getRootNode(vNode.actualNode);
+	const owner = doc.querySelector(
+		`[aria-owns~=${axe.utils.escapeSelector(id)}]`
+	);
+	if (!owner) {
+		return getAriaOwner(vNode.parent);
+	}
+	if (['presentation', 'none'].includes(aria.getRole(owner))) {
+		return;
```
Comment (WilcoFiers) [inline lib/commons/aria/get-owner-virtual.js]: aria-owns is a global attribute, role=none should get ignored. We should probably test that to be sure, but I suspect this if statement needs to be remove.

File: lib/commons/aria/get-owner-virtual.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	if (!id) {
+		return getAriaOwner(vNode.parent);
+	}
+
+	const doc = dom.getRootNode(vNode.actualNode);
+	const owner = doc.querySelector(
+		`[aria-owns~=${axe.utils.escapeSelector(id)}]`
+	);
+	if (!owner) {
+		return getAriaOwner(vNode.parent);
```
Comment (WilcoFiers) [inline lib/commons/aria/get-owner-virtual.js]: As above, I aria-owns doesn't cause all descendants to change owner, just the element that's directly referenced.

File: lib/commons/aria/get-owner-virtual.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
+/* global aria, dom */
+
+/**
+ * Get owner element of a provided virtualNode
+ * @param {Object} virtualNode virtual node
+ * @returns {HTMLElement | undefined}
+ */
+aria.getOwnerVirtual = function getOwnerVirtual(virtualNode) {
+	return getAriaOwner(virtualNode) || getOwner(virtualNode);
```
Comment (WilcoFiers) [inline lib/commons/aria/get-owner-virtual.js]: Before doing this, can you test that the thing doesn't have role=presentation or none. Those elements can't be owned.

File: lib/commons/aria/get-owner-virtual.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
+/* global aria, dom */
+
+/**
+ * Get owner element of a provided virtualNode
+ * @param {Object} virtualNode virtual node
+ * @returns {HTMLElement | undefined}
```
Comment (WilcoFiers) [inline lib/commons/aria/get-owner-virtual.js]: This needs to return virtual nodes. We have to make sure this works on SerialVirtualNode too.

File: lib/commons/aria/get-owner-virtual.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	if (!vNode) {
+		return;
+	}
+
+	const id = vNode.attr('id');
+	if (!id) {
+		return getAriaOwner(vNode.parent);
+	}
+
+	const doc = dom.getRootNode(vNode.actualNode);
```
Comment (WilcoFiers) [inline lib/commons/aria/get-owner-virtual.js]: Please check if vNode.actualNode exists first.

File: lib/checks/aria/required-parent.js [hunk 10]
```diff
@@ -1,87 +1,41 @@
...
-				getSelector(requiredContext[index])
-			)
-		) {
-			//if one matches, it passes
-			return null;
-		} else {
-			missing.push(requiredContext[index]);
-		}
+	const role = aria.getRole(elm);
+	if (['presentation', 'none', null].includes(aria.getRole(elm))) {
```
Comment (WilcoFiers) [inline lib/checks/aria/required-parent.js]: You've already tested for presentation/none in getOwnerVirtual, why are you doing this again? Also, why are you ignoring null? Firefox doesn't ignore null roles, so I don't think we should either. It'd be good to add an option to this check to make it configurable whether or not null is ignored, but that should probably be handled by getOwnerVirtual. Can you either add that or open an issue for it?

File: lib/commons/aria/get-owner-virtual.js (hunk 7 above)
Comment (WilcoFiers) [inline lib/commons/aria/get-owner-virtual.js]: Also thinking we should add options to this, allowing for the following:
- noAriaOwns
- ignoreNullRoleOwner

These could be passed in from the check options. Either add these now or open an issue for it.

File: lib/checks/aria/required-parent.js [hunk 11]
```diff
@@ -1,87 +1,41 @@
...
 }
 
-function getMissingContext(virtualNode, requiredContext, includeElement) {
-	var index,
-		length,
-		role = virtualNode.actualNode.getAttribute('role'),
-		missing = [];
+const owner = aria.getOwnerVirtual(virtualNode);
+if (!owner) {
+	return undefined;
```
Comment (WilcoFiers) [inline lib/checks/aria/required-parent.js]: add a message to the check, and add tests.

File: test/integration/rules/aria-required-parent/aria-required-parent.html [hunk 12]
```diff
@@ -1,16 +1,51 @@
...
 <div role="menu" id="pass6" aria-owns="fail3"></div>
+
 <div role="listitem" id="fail3"></div>
+
 <div role="list" id="pass7" aria-owns="parent"></div>
-<div id="parent"><div role="listitem" id="pass8"></div></div>
-<div role="grid" id="pass9"><div role="rowgroup" id="pass10">Item 1</div></div>
+
+<div id="parent">
+	<div role="listitem" id="pass8"></div>
```
Comment (WilcoFiers) [inline test/integration/rules/aria-required-parent/aria-required-parent.html]: I don't think this should pass. aria-owns needs to point directly to the owned element, not one of its descendants. Can you test this to make sure?

File: test/commons/aria/get-owner-virtual.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+		axe._tree = undefined;
+	});
+
+	it('returns owner element - `div` which is a direct parent (not `aria-owns`)', function() {
+		var vNode = queryFixture(
+			'<div><p role="listitem" id="target">Nothing here.</p></div>'
+		);
+
+		var owner = aria.getOwnerVirtual(vNode);
+		assert.equal(owner.nodeName.toUpperCase(), 'DIV');
```
Comment (WilcoFiers) [inline test/commons/aria/get-owner-virtual.js]: There can be any number of divs on this page. This test doesn't prove what you're trying to prove.

File: test/commons/aria/get-owner-virtual.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+	it('returns owner element which references via `aria-owns`', function() {
+		var vNode = queryFixture(
+			'<div role="list"><div aria-owns="parent"></div></div>' +
+				'<div id="parent"><p role="listitem" id="target">Nothing here.</p></div>'
+		);
+
+		var owner = aria.getOwnerVirtual(vNode);
+		assert.equal(owner.nodeName.toUpperCase(), 'DIV');
+		assert.equal(owner.getAttribute('aria-owns'), 'parent');
```
Comment (WilcoFiers) [inline test/commons/aria/get-owner-virtual.js]: the id=parent element should be the owner here, not the element with aria-owns.

**Comment by jkodu** [comment]
Given the change to es modules, it may be worth to re-do this after the conversion to es modules, to avoid a lot of conflict resolution.

**Pull request closed without merging**